burstphoto/align/align.swift

```diff
@@ -73,10 +74,12 @@ func align_texture(_ ref_pyramid: [MTLTexture], _ comp_texture: MTLTexture, _ do
 }
 
 
-func avg_pool(_ input_texture: MTLTexture, _ scale: Int, _ normalization: Bool, _ color_factors3: Array<Double>) -> MTLTexture {
+func avg_pool(_ input_texture: MTLTexture, _ scale: Int, _ black_level_mean: Double, _ normalization: Bool, _ color_factors3: Array<Double>) -> MTLTexture {
```

Comment:
Currently, wherever required, a color-channel dependent black level is used, i.e. 4 black level values are available. I agree that there might be cameras, where even row/column-specific black levels might be needed. For some functions however, this level of detail is not required as one global value for shifting the pixel values close to zero is needed. For the avg_pool() function, this is needed as the input_texture has 32 bit float precision and the output texture has only 16 bit float precision. Just think of a black value of 512 and a pixel value of 517. Between 512 - 1024, the float16 format has a precision of 512/1024 = 0.5. When you now subtract 512 (a slightly off value would not really change anything), you are in the range 4-8, where you have a precision of 4/1024 = 0.00390625, which is approx. 3 orders of magnitude better. The only purpose of the subtraction of the black level in this function is a considerable improvement of the precision of pixel values in the shadows, which ever so slightly improves the quality of alignment.

Reply:
Okay, I understand that part. The only reason this is needed is because the black level is not subtracted from the textures at this point. However, if `image_url_to_texture` returned a texture with the black level already subtracted. Then this would not be required anymore. Correct?

I ask because I would prefer to support the more complicated cameras (with row/column dependent black level) by having `image_url_to_texture` to return a raw with the black level already subtracted. The alternative would be carrying around a 2nd 16bit (I believe) float texture.

Reply:
Ideally, the black level would be subtracted once right at the beginning of the pipeline. At the export stage, it would then be added to ensure file compatibility (options here can be discussed). Unfortunately, it is not that simple. There is at least one function where a multiplication of pixel values is performed BEFORE the black level subtraction. In addition, NR reduction strength is tuned to pixel intensities before black level subtraction. So in any way, some operations with black level values will be required at different stages of the pipeline. Now the question is, is it more consistent to subtract at the beginning and add if required during the pipeline execution or is it better to let it as it currently is and subtract if required during pipeline execution? I fear the scope of this change is larger than expected. Another solution would be to extend the black level information and make subtraction or addition easier to execute so that more complicated black level patterns can be supported.

Reply:
Yes, I agree that revamping of black level handling will be its own undertaking, I wanted to flag it here since this PR introduces further dependencies on the black level into the pipeline. I'll open a separate issue to talk about this and track it in there.